tensorboard/plugins/mesh_visualizer/mesh_plugin.py

```diff
@@ -0,0 +1,253 @@
+
+    This method is called by TensorBoard when retrieving all the
+    routes offered by the plugin.
+    Returns:
+      A dictionary mapping URL path to route that handles it.
+    """
+    # Note that the methods handling routes are decorated with
+    # @wrappers.Request.application.
+    return {
+        '/tags': self._serve_tags,
+        '/meshes': self._serve_mesh_metadata,
+        '/data': self._serve_mesh_data
+    }
+
+  def is_active(self):
+    """Determines whether this plugin is active.
+
+    This plugin is only active if TensorBoard sampled any summaries
+    relevant to the mesh plugin.
+    Returns:
+      Whether this plugin is active.
+    """
+    all_runs = self._multiplexer.PluginRunToTagToContent(
+        MeshPlugin.plugin_name)
+
+    # The plugin is active if any of the runs has a tag relevant
+    # to the plugin.
+    return bool(self._multiplexer and any(six.itervalues(all_runs)))
+
+  def _get_sample(self, tensor_event, sample):
+    """Returns a single sample from a batch of samples."""
+    data = tensor_util.make_ndarray(tensor_event.tensor_proto)
+    return data[sample].tolist()
+
+  def _get_tensor_metadata(self, event, content_type, data_shape, config):
+    """Converts a TensorEvent into a JSON-compatible response.
+
+    Args:
+      event: TensorEvent object containing data in proto format.
+      content_type: enum plugin_data_pb2.MeshPluginData.ContentType value,
+        representing content type in TensorEvent.
+      data_shape: list of dimensions sizes of the tensor.
+      config: rendering scene configuration as dictionary.
+    Returns:
+      Dictionary of transformed metadata.
+    """
+    return {
+        'wall_time': event.wall_time,
+        'step': event.step,
+        'content_type': content_type,
+        'config': config,
+        'data_shape': list(data_shape)
+    }
+
+  def _get_tensor_data(self, event, sample):
+    """Convert a TensorEvent into a JSON-compatible response."""
+    data = self._get_sample(event, sample)
+    return data
+
+  def _collect_tensor_events(self, request):
+    """Collects list of tensor events based on request."""
+    run = request.args.get('run')
+    tag = request.args.get('tag')
+
+    # TODO(b/128995556): investigate why this additional metadata mapping is
+    # necessary, it must have something todo with the lifecycle of the request.
+    # Make sure we populate tags mapping structures.
+    self.prepare_metadata()
+
+    # We fetch all the tensor events that contain tag.
+    tensor_events = []  # List of tuples (meta, tensor).
+    for instance_tag in self._tag_to_instance_tags[tag]:
+      tensors = self._multiplexer.Tensors(run, instance_tag)
+      meta = self._instance_tag_to_metadata[instance_tag]
+      tensor_events += [(meta, tensor) for tensor in tensors]
+
+    # Make sure tensors sorted by timestamp in ascending order.
+    tensor_events = sorted(
+        tensor_events, key=lambda tensor_data: tensor_data[1].wall_time)
```

Comment:
Consider `key=lambda (_, event): event.wall_time`.

Reply:
It fails with error in PY3 mode:
./tensorboard/plugins/mesh_visualizer/mesh_plugin.py:197:35: E999 SyntaxError: invalid syntax
        tensor_events, key=lambda (_, event): event.wall_time)
                                  ^
1     E999 SyntaxError: invalid syntax

Reply:
Ugh, that’s unfortunate, to say the least. Good to know; thanks.